File: src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs
// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved. See License.md in the project root for license information.

using System;
using System.Diagnostics.CodeAnalysis;
using System.IO;
using System.Threading;
using Microsoft.AspNet.SignalR.Client.Http;
using Microsoft.AspNet.SignalR.Client.Infrastructure;
using Microsoft.AspNet.SignalR.Client.Transports.ServerSentEvents;
using Microsoft.AspNet.SignalR.Infrastructure;

namespace Microsoft.AspNet.SignalR.Client.Transports
{
    public class ServerSentEventsTransport : HttpBasedTransport
    {
        private IRequest _request;
        private bool _stop;

        public ServerSentEventsTransport()
            : this(new DefaultHttpClient())
        {
        }

        public ServerSentEventsTransport(IHttpClient httpClient)
            : base(httpClient, "serverSentEvents")
        {
            ReconnectDelay = TimeSpan.FromSeconds(2);
        }

        /// <summary>
        /// Indicates whether or not the transport supports keep alive
        /// </summary>
        public override bool SupportsKeepAlive
        {
            get
            {
                return true;
            }
        }

        /// <summary>
        /// The time to wait after a connection drops to try reconnecting.
        /// </summary>
        public TimeSpan ReconnectDelay { get; set; }

        protected override void OnStart(IConnection connection, string connectionData, CancellationToken disconnectToken)
        {
            OpenConnection(connection, connectionData, disconnectToken, reconnecting: false);
        }

        protected override void OnStartFailed()
        {
            // if the transport failed to start we want to stop it silently.
            _stop = true;

            _request.Abort();
        }

        private void Reconnect(IConnection connection, string data, CancellationToken disconnectToken)
        {
            // Need to verify before the task delay occurs because an application sleep could occur during the delayed duration.
            if (!TransportHelper.VerifyLastActive(connection))
            {
                return;
            }

            // Wait for a bit before reconnecting
            TaskAsyncHelper.Delay(ReconnectDelay).Then(() =>
            {
                if (!TransportHelper.VerifyLastActive(connection))
                {
                    return;
                }

                // FIX: Race if Connection is stopped and completely restarted between checking the token and calling
                //      connection.EnsureReconnecting()
                if (!disconnectToken.IsCancellationRequested && connection.EnsureReconnecting())
                {
                    // Now attempt a reconnect
                    OpenConnection(connection, data, disconnectToken, reconnecting: true);
                }
            });
        }

        [SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity", Justification = "We will refactor later.")]
        [SuppressMessage("Microsoft.Maintainability", "CA1506:AvoidExcessiveClassCoupling", Justification = "We will refactor later.")]
        internal void OpenConnection(IConnection connection, string data, CancellationToken disconnectToken, bool reconnecting)
        {
            // If we're reconnecting add /connect to the url
            var url = reconnecting
                ? UrlBuilder.BuildReconnect(connection, Name, data)
                : UrlBuilder.BuildConnect(connection, Name, data);

            connection.Trace(TraceLevels.Events, "SSE: GET {0}", url);

            var getTask = HttpClient.Get(url, req =>
            {
                _request = req;
                _request.Accept = "text/event-stream";

                connection.PrepareRequest(_request);

            }, isLongRunning: true);

            var requestCancellationRegistration = disconnectToken.SafeRegister(state =>
            {
                _stop = true;

                // This will no-op if the request is already finished.
                ((IRequest)state).Abort();
            }, _request);
            
            getTask.ContinueWith(task =>
            {
                if (task.IsFaulted || task.IsCanceled)
                {
                    var exception = task.IsCanceled
                        ? new OperationCanceledException(Resources.Error_TaskCancelledException)
                        : task.Exception.Unwrap(); 

                    if (!reconnecting)
                    {
                        TransportFailed(exception);
                    }
                    else if (!_stop)
                    {
                        // Only raise the error event if we failed to reconnect
                        connection.OnError(exception);

                        Reconnect(connection, data, disconnectToken);
                    }

                    requestCancellationRegistration.Dispose();
                }
                else
                {
                    // If the disconnect token is canceled the response to the task doesn't matter.
                    if (disconnectToken.IsCancellationRequested)
                    {
                        return;
                    }

                    var response = task.Result;
                    Stream stream = response.GetStream();

                    var eventSource = new EventSourceStreamReader(connection, stream);

                    eventSource.Opened = () =>
                    {
                        // This will noop if we're not in the reconnecting state
                        if (connection.ChangeState(ConnectionState.Reconnecting, ConnectionState.Connected))
                        {
                            // Raise the reconnect event if the connection comes back up
                            connection.OnReconnected();
                        }
                    };

                    eventSource.Message = sseEvent =>
                    {
                        if (sseEvent.EventType == EventType.Data &&
                            !sseEvent.Data.Equals("initialized", StringComparison.OrdinalIgnoreCase))
                        {
                            ProcessResponse(connection, sseEvent.Data);
                        }
                    };

                    eventSource.Closed = exception =>
                    {
                        if (exception != null)
                        {
                            // Check if the request is aborted
                            if (!ExceptionHelper.IsRequestAborted(exception))
                            {
                                // Don't raise exceptions if the request was aborted (connection was stopped).
                                connection.OnError(exception);
                            }
                        }

                        requestCancellationRegistration.Dispose();
                        response.Dispose();

                        if (_stop)
                        {
                            AbortHandler.CompleteAbort();
                        }
                        else if (AbortHandler.TryCompleteAbort())
                        {
                            // Abort() was called, so don't reconnect
                        }
                        else
                        {
                            Reconnect(connection, data, disconnectToken);
                        }
                    };

                    eventSource.Start();
                }
            });
        }

        public override void LostConnection(IConnection connection)
        {
            if (_request != null)
            {
                _request.Abort();
            }
        }
    }
}

File: src/Microsoft.AspNet.SignalR.Client.Cpp/ServerSentEventsTransport.cpp
//Copyright (c) Microsoft Corporation
//
//All rights reserved.
//
//THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY, OR NON-INFRINGEMENT.

#include "ServerSentEventsTransport.h"

namespace MicrosoftAspNetSignalRClientCpp
{

ServerSentEventsTransport::ServerSentEventsTransport(shared_ptr<IHttpClient> httpClient) : 
    HttpBasedTransport(httpClient, U("serverSentEvents"))
{
    mConnectionTimeout = seconds(5);
    mReconnectDelay = seconds(2);
}

ServerSentEventsTransport::~ServerSentEventsTransport()
{
    int count = pInitializeHandler.use_count();
}

void ServerSentEventsTransport::OnAbort()
{
    // need to clear all the function<> variables to prevent circular referencing
    pEventSource->SetOpenedCallback([](){});
    pEventSource->SetClosedCallback([](exception&){});
    pEventSource->SetDataCallback([](shared_ptr<char>){});
    pEventSource->SetMessageCallback([](shared_ptr<SseEvent>){});
}

void ServerSentEventsTransport::OnStart(shared_ptr<Connection> connection, string_t data, pplx::cancellation_token disconnectToken,  shared_ptr<TransportInitializationHandler> initializeHandler)
{    
    if (initializeHandler == nullptr)
    {
        throw exception("ArgumentNullException: initializeHandler");
    }

    initializeHandler->SetOnFailureCallback([this]()
    {
        mStop = true;
        pRequest->Abort();
    });

    pInitializeHandler = initializeHandler;

    function<void()> successCallback = [this]()
    {
        pInitializeHandler->Success();
        pInitializeHandler.reset();
    };
    function<void(exception&)> errorCallback = [this](exception& ex)
    {
        pInitializeHandler->Fail(ex);
        pInitializeHandler.reset();
    };

    OpenConnection(connection, data, disconnectToken, successCallback, errorCallback);
}

void ServerSentEventsTransport::Reconnect(shared_ptr<Connection> connection, string_t data, pplx::cancellation_token disconnectToken)
{
    TaskAsyncHelper::Delay(mReconnectDelay, disconnectToken).then([this, connection, disconnectToken, data]()
    {
        if (disconnectToken.is_canceled() && connection->EnsureReconnecting())
        {
            OpenConnection(connection, data, disconnectToken, nullptr, nullptr);
        }
    });
}

void ServerSentEventsTransport::OpenConnection(shared_ptr<Connection> connection, string_t data, pplx::cancellation_token disconnectToken, function<void()> initializeCallback, function<void(exception)> errorCallback)
{
    bool reconnecting = initializeCallback == nullptr;
    pCallbackInvoker = unique_ptr<ThreadSafeInvoker>(new ThreadSafeInvoker());
    function<void()> initializeInvoke = [this, initializeCallback]()
    {
        pCallbackInvoker->Invoke(initializeCallback);
    };

    string_t uri = connection->GetUri() + (reconnecting ? U("reconnect") : U("connect")) + GetReceiveQueryString(connection, data);

    wstringstream ss;
    ss << "SSE: GET " << uri;
    connection->Trace(TraceLevel::Events, ss.str());

    GetHttpClient()->Get(uri, [this, connection](shared_ptr<HttpRequestWrapper> request)
    {
        pRequest = request;
        connection->PrepareRequest(request);
    }).then([this, connection, data, disconnectToken, errorCallback, initializeInvoke, reconnecting](pplx::task<http_response> connectRequest) 
    {
        http_response response;
        exception ex;
        TaskStatus status = TaskAsyncHelper::RunTaskToCompletion<http_response>(connectRequest, response, ex);
        
        if (status == TaskStatus::TaskFaulted)
        {
            if (!ExceptionHelper::IsRequestAborted(ex))
            {
                if (errorCallback != nullptr)
                {
                    pCallbackInvoker->Invoke<function<void(exception&)>, exception&>([](function<void(exception&)> cb, exception& ex)
                    {
                        cb(ex);        
                    }, errorCallback, ex);
                }
                else if (reconnecting)
                {
                    // raise error event if failed to reconnect
                    connection->OnError(ex);
                    Reconnect(connection, data, disconnectToken);
                }
            }
            {
                lock_guard<mutex> lock(mDeregisterRequestCancellationLock);
                DeregisterRequestCancellation();
                DeregisterRequestCancellation = [](){};
            }
        }
        else if (status == TaskStatus::TaskCanceled)
        {
            return;
        }
        else
        {
            pEventSource = unique_ptr<EventSourceStreamReader>(new EventSourceStreamReader(connection, response.body()));

            mStop = false;
            bool stop = false;
        
            // what is CancellationTokenExtensions.SafeRegister?
            disconnectToken.register_callback<function<void()>>([this, stop]()
            {
                mStop = true;
                pRequest->Abort();
            });

            pEventSource->SetOpenedCallback([connection]()
            {
                if (connection->ChangeState(ConnectionState::Reconnecting, ConnectionState::Connected))
                {
                    connection->OnReconnected();
                }
            });

            pEventSource->SetMessageCallback([connection, this, initializeInvoke, stop](shared_ptr<SseEvent> sseEvent) 
            {
                if (sseEvent->GetType() == EventType::Data)
                {
                    if (StringHelper::EqualsIgnoreCase(sseEvent->GetData(), string_t(U("initialized"))))
                    {
                        return;
                    }

                    bool timedOut, disconnected;

                    TransportHelper::ProcessResponse(connection, sseEvent->GetData(), &timedOut, &disconnected, initializeInvoke);
                    disconnected = false;

                    if (disconnected)
                    {
                        mStop = true;
                        connection->Disconnect();
                    }
                }
            });

            pEventSource->SetClosedCallback([connection, this, data, disconnectToken, stop](exception& ex)
            {
                if (!ExceptionHelper::IsNull(ex))
                {
                    // Check if the request is aborted
                    if (!ExceptionHelper::IsRequestAborted(ex))
                    {
                        // Don't raise exceptions if the request was aborted (connection was stopped).
                        connection->OnError(ex);
                    }
                }

                {
                    lock_guard<mutex> lock(mDeregisterRequestCancellationLock);
                    DeregisterRequestCancellation();
                    DeregisterRequestCancellation = [](){};
                }

                if (mStop)
                {
                    GetAbortHandler()->CompleteAbort();
                }
                else if (GetAbortHandler()->TryCompleteAbort())
                {
                    // Abort() was called, so don't reconnect
                }
                else
                {
                    Reconnect(connection, data, disconnectToken);
                }
            });

            pEventSource->Start();
        }
    });

    auto requestCancellationRegistration = disconnectToken.register_callback([this, errorCallback]()
    {
        if (pRequest != nullptr)
        {
            pRequest->Abort();
        }

        if (errorCallback != nullptr)
        {
            pCallbackInvoker->Invoke<function<void(exception&)>>([](function<void(exception&)> cb)
            {
                cb(exception("Operation Canceled Exception: The connection was stopped before it could be started."));
            }, errorCallback);
        }
    });

    {
        lock_guard<mutex> lock(mDeregisterRequestCancellationLock);
        DeregisterRequestCancellation = [disconnectToken, requestCancellationRegistration]()
        {
            disconnectToken.deregister_callback(requestCancellationRegistration);
        };
    }

    if (errorCallback != nullptr)
    {
        // the memory will eventually be reclaimed after 5 seconds (default) so it's not a leak here
        TaskAsyncHelper::Delay(mConnectionTimeout, disconnectToken).then([this, errorCallback]()
        {
            pCallbackInvoker->Invoke<function<void(exception)>>([this](function<void(exception)> errorCallback)
            {
                // Abort the request before cancelling
                pRequest->Abort();

                // Connection timeout occurred
                errorCallback(exception("TimeoutException"));
            }, errorCallback);
        });
    }
}

void ServerSentEventsTransport::LostConnection(shared_ptr<Connection> connection)
{
    if (pRequest != nullptr)
    {
        pRequest->Abort();
    }
}

bool ServerSentEventsTransport::SupportsKeepAlive()
{
    return true;
}

} // namespace MicrosoftAspNetSignalRClientCpp